pkg/ddl/ddl_api.go

```diff
@@ -4603,6 +4603,41 @@ func (d *ddl) AlterTablePartitioning(ctx sessionctx.Context, ident ast.Ident, sp
 	}
 	newPartInfo := newMeta.Partition
 
+	// Only non-partition table to partition table need to check global index.
+	if piOld.Type == model.PartitionTypeNone {
```

Comment:
```suggestion
```
I think it needs to be checked for all types, since this is still accepted:
```mysql
create table t4 (a int primary key, b int) partition by hash(a) partitions 3;
alter table t4 partition by hash(b) partitions 3;
```

Reply:
Yes, you are right. Updated https://github.com/pingcap/tidb/pull/51635/commits/6ab2ba122c80b701a048f6b426c346893e3c1a5f.